sources/cassandra/validations/count.go

```diff
@@ -0,0 +1,436 @@
+		}
+	}
+
+	return cluster.CreateSession()
+}
+
+// connectToSpanner establishes a connection to the Spanner database
+func connectToSpanner(ctx context.Context, uri string) (*spanner.Client, error) {
+	return spanner.NewClient(ctx, uri)
+}
+
+// countSpannerRows counts the total number of rows in a Spanner table
+func countSpannerRows(ctx context.Context, client *spanner.Client, table string) (int64, error) {
+	stmt := spanner.Statement{
+		SQL: fmt.Sprintf("SELECT COUNT(*) FROM %s", table),
+	}
+
+	iter := client.Single().Query(ctx, stmt)
+	defer iter.Stop()
+
+	row, err := iter.Next()
+	if err == iterator.Done {
+		return 0, fmt.Errorf("no results returned from count query")
+	}
+	if err != nil {
+		return 0, fmt.Errorf("error executing count query: %w", err)
+	}
+
+	var count int64
+	if err := row.Columns(&count); err != nil {
+		return 0, fmt.Errorf("error scanning count result: %w", err)
+	}
+
+	return count, nil
+}
+
+// getClusterTokenRanges retrieves and processes token ranges from all nodes in the Cassandra cluster.
+// It combines tokens from both local and peer nodes, deduplicates them, and creates a sorted list of
+// token ranges that cover the entire token ring.
+// TODO: Consider creating configurable number of partitions instead of system ranges.
+func getClusterTokenRanges(session *gocql.Session) ([]TokenRange, error) {
+	localTokens, err := getLocalNodeTokens(session)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get local node tokens: %w", err)
+	}
+
+	peerTokens, err := getPeerNodeTokens(session)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get peer node tokens: %w", err)
+	}
+
+	// 1. Combine and Deduplicate Tokens
+	allTokens := append(localTokens, peerTokens...)
+	uniqueTokens := distinct(allTokens)
+
+	// 2. Map to Big Ints and Sort
+	bigIntTokens := make([]*big.Int, len(uniqueTokens))
+	for i, tokenStr := range uniqueTokens {
+		bigIntToken := new(big.Int)
+		if _, ok := bigIntToken.SetString(tokenStr, 10); !ok {
+			return nil, fmt.Errorf("invalid token string: %s", tokenStr)
+		}
+		bigIntTokens[i] = bigIntToken
+	}
+	sort.Slice(bigIntTokens, func(i, j int) bool {
+		return bigIntTokens[i].Cmp(bigIntTokens[j]) < 0
+	})
+
+	// 3. Create Token Ranges
+	tokenRanges := make([]TokenRange, len(bigIntTokens)+1)
+	// First range
+	tokenRanges[0] = TokenRange{Start: "nil", End: bigIntTokens[0].String()}
+
+	// Middle ranges
+	for i := 1; i < len(bigIntTokens); i++ {
+		tokenRanges[i].Start = bigIntTokens[i-1].String()
+		tokenRanges[i].End = bigIntTokens[i].String()
+
+	}
+
+	// Last range
+	lastIndex := len(tokenRanges) - 1
+	tokenRanges[lastIndex] = TokenRange{Start: bigIntTokens[lastIndex-1].String(), End: "nil"}
+
+	return tokenRanges, nil
+}
+
+// distinct takes a slice of strings and returns a new slice containing only unique elements,
+// removing any duplicates from the input slice.
+func distinct(tokens []string) []string {
+	seen := make(map[string]bool)
+	unique := []string{}
```

Comment:
nit: logic seems a little roundabout. it would probably just be easier to dump all tokens into a map and convert it into a list. The duplicates would get dropped anyway.

Reply:
Agreed that could be an optimisation. Added a todo since this wont be a performance bottleneck as its a one time thing.